src/common/libfluxutil/conf_bootstrap.c

```diff
@@ -0,0 +1,464 @@
+    hostlist_destroy (hl);
+    return -1;
+}
+
+/* Parse and validate the hosts array
+ * Assign the expanded hosts array to binfo->xhosts.
+ * Return -1 on failure, instance size on success.
+ */
+static int parse_hosts (json_t *hosts,
+                        const char *hostname,
+                        struct bootstrap_info *binfo,
+                        flux_error_t *errp)
+{
+    size_t index;
+    json_t *entry;
+    json_t *hobj = NULL;
+    flux_error_t error;
+
+    if (hosts && !json_is_array (hosts))
+        return errprintf (errp, "must be array type");
+    if (!(binfo->xhosts = json_array ()))
+        goto nomem;
+    // missing or empty hosts array is allowed - assume singleton
+    if (!hosts || json_array_size (hosts) == 0) {
+        if (!(entry = json_pack ("{s:s}", "host", hostname))
+            || json_array_append_new (binfo->xhosts, entry) < 0) {
+            json_decref (entry);
+            goto nomem;
+        }
+        return 1;
+    }
+    if (!(hobj = json_object ())) // just a tool for deduplication
+        goto nomem;
+    // first pass - expand hostlists and combine duplicates => binfo->xhosts
+    json_array_foreach (hosts, index, entry) {
+        if (dedup_hosts_entry (binfo, entry, hobj, &error) < 0) {
+            errprintf (errp, "invalid hosts[%zu]: %s", index, error.text);
+            goto error;
+        }
+    }
+    // second pass - apply defaults and validate
+    json_array_foreach (binfo->xhosts, index, entry) {
+        if (parse_hosts_entry (binfo, entry, hobj, &error) < 0) {
+            errprintf (errp, "invalid hosts[%zu]: %s", index, error.text);
+            goto error;
+        }
+    }
+    // ensure local hostname is is present
+    if (!json_object_get (hobj, hostname)) {
+        errprintf (errp, "%s not found in hosts", hostname);
+        goto error;
+    }
+    json_decref (hobj);
+    return json_array_size (binfo->xhosts);
+nomem:
+    errprintf (errp, "out of memory");
+error:
+    json_decref (hobj);
+    json_decref (binfo->xhosts);
+    binfo->xhosts = NULL;
+    return -1;
+}
+
+/* Check that TCP port number is usable.
+ */
+static int validate_port (int port, flux_error_t *errp)
+{
+    if (port != -1) {
+        if (port < 0 || port > 65535)
+            return errprintf (errp, "must be in the range of 0-65535");
+    }
+    return 0;
+}
+
+static int validate_curve_cert (const char *path, int size, flux_error_t *errp)
+{
+    if (size > 1 && !path)
+        return errprintf (errp, "curve_cert must be defined for size > 1");
+    return 0;
+}
+
+int conf_bootstrap_parse (const flux_conf_t *conf,
+                          const char *hostname,
+                          bool *enable_ipv6,
+                          const char **curve_cert,
+                          json_t **hosts,
+                          flux_error_t *errp)
+{
+    flux_error_t error;
+    json_error_t jerror;
+    json_t *bootstrap = NULL;
+    struct bootstrap_info binfo = { 0 };
```

Comment:
Would it be simpler and more clear to directly initialize `default_port` to `-1`?
```c
    struct bootstrap_info binfo = { .default_port = -1 };
```